### PR TITLE
Allow adding and editing fiat assets

### PR DESCRIPTION
**Pull request by nebolax**
Closes #(issue_number)

## Checklist

- [ ] The PR modified the frontend, and updated the [user guide](https://github.com/rotki/rotki/blob/develop/docs/usage_guide.rst) to reflect the changes.

### CODE REVIEW

File: rotkehlchen/tests/api/test_user_assets.py [hunk 1]
```diff
@@ -339,7 +362,7 @@ def test_editing_user_assets(rotkehlchen_api_server, globaldb):
     )
     # try to add invalid cryptocompare
     bad_id = 'dsadsad'
-    bad_asset = {
+    evm_token_data = {
```
Comment (yabirgb) [inline rotkehlchen/tests/api/test_user_assets.py:365]: why rename it to `evm_token_data`? is not an evm token. Same to similar changes in this test

File: rotkehlchen/serialization/schemas.py [hunk 2]
```diff
@@ -102,24 +97,74 @@ def validate_schema(
...
+
+
+class CryptoAssetFieldsSchema(Schema):
+    started = TimestampField(load_default=None)
+    swapped_for = AssetField(expected_type=CryptoAsset, load_default=None)
+    forked = AssetField(expected_type=CryptoAsset, load_default=None)
+
+
+class CryptoAssetSchema(AssetWithOraclesSchema, CryptoAssetFieldsSchema):
+    asset_type = AssetTypeField(load_default=None)
```
Comment (yabirgb) [inline rotkehlchen/serialization/schemas.py:132]: Could you add docstrings to this functions? So it is easier to know why you have something that is fields and another that has only crypto asset. At first is hard to follow

File: rotkehlchen/serialization/schemas.py [hunk 3]
```diff
@@ -102,24 +97,74 @@ def validate_schema(
...
+            identifier=data['identifier'] if data['identifier'] is not None else str(uuid4()),
+            name=data['name'],
+            symbol=data['symbol'],
+            coingecko=data['coingecko'],
+            cryptocompare=data['cryptocompare'],
+        )
+
+
+class CryptoAssetFieldsSchema(Schema):
+    started = TimestampField(load_default=None)
```
Comment (yabirgb) [inline rotkehlchen/serialization/schemas.py:126]: Why not make it inherit from `AssetWithOraclesFieldsSchema`?

File: rotkehlchen/serialization/schemas.py [hunk 4]
```diff
@@ -102,24 +97,74 @@ def validate_schema(
...
+    swapped_for = AssetField(expected_type=CryptoAsset, load_default=None)
+    forked = AssetField(expected_type=CryptoAsset, load_default=None)
+
+
+class CryptoAssetSchema(AssetWithOraclesSchema, CryptoAssetFieldsSchema):
```
Comment (yabirgb) [inline rotkehlchen/serialization/schemas.py:131]: Why are you using here `AssetWithOraclesSchema` and in `EvmTokenSchema` you are using `AssetWithOraclesFieldsSchema`?

**Review by yabirgb — COMMENTED** [review]
Some questions about the approach taken

File: rotkehlchen/serialization/schemas.py [hunk 5]
```diff
@@ -102,24 +97,74 @@ def validate_schema(
...
+    @validates_schema
+    def validate_schema(
+            self,
+            data: dict[str, Any],
+            **_kwargs: Any,
+    ) -> None:
+        super().validate_schema(data, **_kwargs)
+        if self.expected_asset_type is None and data['asset_type'] is None:
+            raise ValidationError(message='asset_type should be given', field_name='asset_type')
+        if self.expected_asset_type is not None and data['asset_type'] is not None and data['asset_type'] != self.expected_asset_type:  # noqa: E501
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/serialization/schemas.py:157]: No because if an exception is raised, the execution of a function is aborted

File: rotkehlchen/globaldb/handler.py [hunk 6]
```diff
@@ -1031,6 +1031,9 @@ def edit_user_asset(asset: AssetWithOracles) -> None:
                 f'Tried to edit non existing asset with identifier {asset.identifier}',
             ) from e
 
+        if asset.is_evm_token():
+            GlobalDBHandler().edit_evm_token(cast(EvmToken, asset))
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/globaldb/handler.py:1035]: Perhaps yes. Let me ask frontend guys about it

File: rotkehlchen/tests/api/test_user_assets.py (hunk 1 above)
Comment (nebolax) [reply rotkehlchen/tests/api/test_user_assets.py:365]: Thanks for noticing it. It got autorenamed unexpectedly. I meant to rename only one such dict, where there is evm token data, but since they all weer named `bad_asset`, all of them were renamed

File: rotkehlchen/serialization/schemas.py (hunk 4 above)
Comment (nebolax) [reply rotkehlchen/serialization/schemas.py:131]: Explained in a docstring in CryptoAssetFieldsSchema

File: rotkehlchen/serialization/schemas.py [hunk 7]
```diff
@@ -133,7 +184,11 @@ def transform_data(self, data: dict[str, Any], **_kwargs: Any) -> dict[str, Cryp
         return {'crypto_asset': crypto_asset}
 
 
-class EvmTokenSchema(BaseCryptoAssetSchema, RequiredEvmAddressSchema):
+class EvmTokenSchema(
+    AssetWithOraclesSchema,
+    CryptoAssetFieldsSchema,
```
Comment (yabirgb) [inline rotkehlchen/serialization/schemas.py:189]: This thing with the schemas looks confusing and seems that need to be tough a bit better. For example can't `CryptoAssetFieldsSchema` inherit from `AssetWithOraclesSchema` and then you can avoid having `AssetWithOraclesSchema` both in crypto and evm schemas?

File: rotkehlchen/serialization/schemas.py [hunk 8]
```diff
@@ -247,33 +296,65 @@ def make_custom_asset(
...
+            self,
+            identifier_required: bool,
+            disallowed_asset_types: Optional[list[AssetType]] = None,
+            coingecko: Optional['Coingecko'] = None,
+            cryptocompare: Optional['Cryptocompare'] = None,
+            **kwargs: Any,
+    ) -> None:
+        super().__init__(**kwargs)
+        self.identifier_required = identifier_required
+        self.disallowed_asset_types = disallowed_asset_types
```
Comment (yabirgb) [inline rotkehlchen/serialization/schemas.py:316]: Can you document in the docstring this two arguments? I bet that in two weeks I'll need to use those and I won't recall what I'm suppose to use

**Review by yabirgb — COMMENTED** [review]
Also there is a test that failed now and in the previous run. Please check it

File: rotkehlchen/api/v1/resources.py [hunk 9]
```diff
@@ -832,32 +825,11 @@ def put(self, source_identifier: str, target_asset: Asset) -> Response:
...
-        )
 
     @use_kwargs(get_schema, location='json_and_query')
     def get(self, address: Optional[ChecksumEvmAddress], evm_chain: ChainID) -> Response:
         return self.rest_api.get_custom_evm_tokens(address=address, chain_id=evm_chain)
 
-    @require_loggedin_user()
-    @resource_parser.use_kwargs(make_edit_schema, location='json')
-    def put(self, token: EvmToken) -> Response:
-        return self.rest_api.add_custom_evm_token(token=token)
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/api/v1/resources.py:850]: Done

File: rotkehlchen/serialization/schemas.py (hunk 7 above)
Comment (nebolax) [reply rotkehlchen/serialization/schemas.py]: Yes, I agree that it's a bit hard to follow the schemas right now. Rearranged it a bit. Hope it became more readable

File: rotkehlchen/globaldb/handler.py (hunk 6 above)
Comment (LefterisJP) [inline rotkehlchen/globaldb/handler.py:1035]: shouldnt''t you return or something here? Otherwise this will go further down with `asset.is_crypto()` evaluating to `True` and the asset would be edited there too.

Also just lose the cast and use #type: ignore saying its' checked in the if.

File: rotkehlchen/serialization/schemas.py [hunk 10]
```diff
@@ -102,24 +89,79 @@ def validate_schema(
...
+            name=data['name'],
+            symbol=data['symbol'],
+            coingecko=data['coingecko'],
+            cryptocompare=data['cryptocompare'],
+        )
+
+
+class CryptoAssetFieldsSchema(AssetWithOraclesSchema):
+    """
+    Adds ields specific to crypto assets.
```
Comment (LefterisJP) [inline rotkehlchen/serialization/schemas.py:119]:
```suggestion
    Adds fields specific to crypto assets.
```

File: rotkehlchen/serialization/schemas.py [hunk 11]
```diff
@@ -197,6 +235,7 @@ def transform_data(
         given_underlying_tokens = data.pop('underlying_tokens', None)
         chain_id = data.pop('evm_chain')
         data['chain_id'] = chain_id
+        data.pop('identifier')
```
Comment (LefterisJP) [inline rotkehlchen/serialization/schemas.py:238]: Perhaps it would makesense to initialize `EvmToken.initialize()` below by providing the arguments directly and getting rid of this extra logic to prepare data dict to have the exact amount of arguments required. Can you do that?

File: rotkehlchen/serialization/schemas.py [hunk 12]
```diff
@@ -247,33 +285,72 @@ def make_custom_asset(
...
+            self,
+            identifier_required: bool,
+            disallowed_asset_types: Optional[list[AssetType]] = None,
+            coingecko: Optional['Coingecko'] = None,
+            cryptocompare: Optional['Cryptocompare'] = None,
+            **kwargs: Any,
+    ) -> None:
+        """
+        Initializes an asset schema.
+        If identifier_required is True then the identifier field is required.
```
Comment (LefterisJP) [inline rotkehlchen/serialization/schemas.py:305]:
```suggestion
        Initializes an asset schema depending on the given asset type.

        If identifier_required is True then the identifier field is required.
```

File: rotkehlchen/serialization/schemas.py [hunk 13]
```diff
@@ -247,33 +285,72 @@ def make_custom_asset(
...
+                coingecko=self.coingecko_obj,
+                cryptocompare=self.cryptocompare_obj,
+            ).load(data)
         elif asset_type == AssetType.EVM_TOKEN:
-            return EvmTokenSchema().load(data)
+            asset = EvmTokenSchema(
+                coingecko=self.coingecko_obj,
+                cryptocompare=self.cryptocompare_obj,
+            ).load(data)
         else:
```
Comment (LefterisJP) [inline rotkehlchen/serialization/schemas.py:343]:
```suggestion
        else:  # only other case is generic crypto asset
```

File: rotkehlchen/tests/api/test_user_assets.py [hunk 14]
```diff
@@ -90,28 +92,35 @@ def test_add_user_assets(rotkehlchen_api_server, globaldb):
     assert data.coingecko == user_asset2['coingecko']
     assert data.cryptocompare == user_asset2['cryptocompare']
 
-    # try to add a token type/name/symbol combo that exists
-    bad_asset = {
-        'asset_type': 'fiat',
-        'name': 'Euro',
-        'symbol': 'EUR',
+    # Add an evm token with the user asset endpoint
+    new_evm_token_address = make_evm_address()
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_user_assets.py:96]: Shuldn't you also add a test for adding a new fiat asset through this endpoint?

**Review by LefterisJP — COMMENTED** [review]
Left some questions.

Also important question here would be if you have coordinated with what the frontend needs to do and can they do it on a new PR based on this?

**Comment by nebolax** [comment]
> Also important question here would be if you have coordinated with what the frontend needs to do and can they do it on a new PR based on this?

Yes, I coordinated. Frontend guys are fine with the changes.

File: rotkehlchen/api/rest.py [hunk 15]
```diff
@@ -1418,96 +1415,6 @@ def replace_asset(self, source_identifier: str, target_asset: Asset) -> Response
         AssetResolver().assets_cache.remove(source_identifier)
         return api_response(OK_RESULT, status_code=HTTPStatus.OK)
 
-    @staticmethod
-    def get_custom_evm_tokens(
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/api/rest.py:1422]: Yep

File: rotkehlchen/globaldb/handler.py (hunk 6 above)
Comment (nebolax) [reply rotkehlchen/globaldb/handler.py:1035]: Yes, will make sense to return here. Thanks

File: rotkehlchen/tests/api/test_user_assets.py (hunk 14 above)
Comment (nebolax) [reply rotkehlchen/tests/api/test_user_assets.py:96]: There actually already is a test a bit lower. I forgot to rename a variable from `bad_asset`, but otherwise the test is there. 

```python

    # Edit a fiat asset with the user asset endpoint
    bad_asset = {
        'identifier': 'EUR',
        'asset_type': 'fiat',
        'name': 'Edited EUR',
        'symbol': 'NEWEUR',
    }
    response = requests.patch(
        api_url_for(
            rotkehlchen_api_server,
            'allassetsresource',
        ),
        json=bad_asset,
    )
    assert_proper_response(response)
    resolved_eur = A_EUR.resolve_to_fiat_asset()
    assert resolved_eur.name == 'Edited EUR'
    assert resolved_eur.symbol == 'NEWEUR'
```

File: rotkehlchen/chain/evm/decoding/decoder.py [hunk 16]
```diff
@@ -147,7 +147,6 @@ def __init__(
         self.chain_modules_prefix_length = len(self.chain_modules_root)
         self.dbevmtx = DBEvmTx(self.database)
         self.dbevents = DBHistoryEvents(self.database)
-        self.base: BaseDecoderTools
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/decoder.py:150]: thanks for remembering

File: rotkehlchen/tests/api/test_user_assets.py (hunk 14 above)
Comment (LefterisJP) [reply rotkehlchen/tests/api/test_user_assets.py:96]: That is editing a fiat asset and it's still called `bad_asset`. I am asking foir **adding** one

**Review by LefterisJP — APPROVED** [review]
LGTM. Will think what makes sense, but probably merging does and let frontend do their thing later
